### PR TITLE
14.0 mig contract_sale_invoicing

### PR DESCRIPTION
**Pull request by dsolanki-initos**
Depends on PR: #577

### CODE REVIEW

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld can you please review this PR?

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld can you please review this?

**Comment by dsolanki-initos** [comment]
It would be great if somone can review this PR

**Comment by dsolanki-initos** [comment]
@max3903 can you please add the reviews here?

**Comment by dsolanki-initos** [comment]
Can somone please review this PR?

**Comment by dsolanki-initos** [comment]
It would be great if we could add reviews here.

**Review by fshah-initos — APPROVED** [review]
LGTM. Functional Tests and code review.

**Comment by dsolanki-initos** [comment]
It would be great if we could merge this PR.

**Comment by dsolanki-initos** [comment]
It would be great if we could add the reviews and merge the PR

**Comment by dsolanki-initos** [comment]
@flotho It would be great if we could add the reviews

**Comment by dsolanki-initos** [comment]
It would be great if we could get the reviews

**Comment by dsolanki-initos** [comment]
It would be great if we could get the reviews

**Comment by dsolanki-initos** [comment]
It would be great if we could get the reviews

**Comment by dsolanki-initos** [comment]
It would be great if we could get the reviews

**Comment by dsolanki-initos** [comment]
It would be great if we could get the reviews

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld 
It would be great if you can add reviews here

**Comment by dsolanki-initos** [comment]
It would be great if we could get the reviews

**Comment by dsolanki-initos** [comment]
@carlos-domatix , @be-cloud-js   It would be great if you both could add the reviews.

**Comment by dsolanki-initos** [comment]
It would be nice if we get reviews here

**Comment by dsolanki-initos** [comment]
It would be nice if we get reviews here

**Comment by dsolanki-initos** [comment]
@rafaelbn  and  @ThomasBinsfeld It would be great if you could add the reviews

File: contract_sale_invoicing/models/contract.py [hunk 1]
```diff
@@ -13,9 +14,8 @@ class ContractContract(models.Model):
         "in contract invoice creation.",
     )
 
-    @api.multi
     def _recurring_create_invoice(self, date_ref=False):
-        invoices = super()._recurring_create_invoice(date_ref)
+        invoices = super(ContractContract, self)._recurring_create_invoice(date_ref)
```
Comment (ThomasBinsfeld) [inline contract_sale_invoicing/models/contract.py]: Keep simplified syntax. with `super()`.

File: contract_sale_invoicing/models/contract.py [hunk 2]
```diff
@@ -28,15 +28,10 @@ def _recurring_create_invoice(self, date_ref=False):
                         contract.partner_id.commercial_partner_id.ids,
                     ),
                     ("invoice_status", "=", "to invoice"),
-                    (
-                        "date_order",
-                        "<=",
-                        "{} 23:59:59".format(contract.recurring_next_date),
-                    ),
+                    ("date_order", ">=", contract.recurring_next_date),
```
Comment (ThomasBinsfeld) [inline contract_sale_invoicing/models/contract.py]: Why this change? <= became >=

File: contract_sale_invoicing/tests/test_contract_sale_invoicing.py [hunk 3]
```diff
@@ -31,7 +32,7 @@ def setUpClass(cls):
                 ],
                 "pricelist_id": cls.partner.property_product_pricelist.id,
                 "analytic_account_id": cls.contract.group_id.id,
-                "date_order": "2016-02-15",
+                "date_order": "2019-02-15",
```
Comment (ThomasBinsfeld) [inline contract_sale_invoicing/tests/test_contract_sale_invoicing.py]: Unit tests should be changed as less as possible during migration. Please revert this change if possible.

**Review by ThomasBinsfeld — CHANGES_REQUESTED** [review]

File: contract_sale_invoicing/models/contract.py (hunk 2 above)
Comment (dsolanki-initos) [reply contract_sale_invoicing/models/contract.py]: To avoid this type of issue :
FAIL: TestContractSaleInvoicing.test_sale_invoicing
Traceback (most recent call last):
  File "/home/travis/build/OCA/contract/setup/contract_sale_invoicing/odoo/addons/contract_sale_invoicing/tests/test_contract_sale_invoicing.py", line 49, in test_sale_invoicing
    self.assertEqual(self.sale_order.invoice_status, "invoiced")
AssertionError: 'to invoice' != 'invoiced'
 to invoice
? ---
invoiced
?

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld  changes are done would be nice if you re-check it.

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld changes are done would be nice if you re-check it and give approval.

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld changes are done would be nice if you re-check it and give approval.

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld changes are done would be nice if you re-check it and give approval.

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld changes are done would be nice if you re-check it and give approval.

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld It would be nice if you give approval to this pull so we can merge this ASAP.

File: contract_sale_invoicing/models/contract.py (hunk 2 above)
Comment (ThomasBinsfeld) [reply contract_sale_invoicing/models/contract.py]: But that can't be correct. This method won't invoice past sale orders now. You can't change the entire behavior because a unit test fails. The issue must be somewhere else.

File: contract_sale_invoicing/models/contract.py [hunk 4]
```diff
@@ -28,15 +28,10 @@ def _recurring_create_invoice(self, date_ref=False):
...
                 ]
             )
             if sales:
-                invoice_ids = sales.action_invoice_create()
-                invoices |= self.env["account.invoice"].browse(invoice_ids)[:1]
-
-        return invoices
+                invoice_ids = sales._create_invoices()
+                invoices |= invoice_ids
+        return invoices[:1]
```
Comment (ThomasBinsfeld) [inline contract_sale_invoicing/models/contract.py]: Why returning only one record?

**Review by ThomasBinsfeld — CHANGES_REQUESTED** [review]

File: contract_sale_invoicing/models/contract.py (hunk 4 above)
Comment (dsolanki-initos) [reply contract_sale_invoicing/models/contract.py]: @ThomasBinsfeld  if this way is incorrect then could you please give a better suggestion.

Comment (dsolanki-initos) [reply contract_sale_invoicing/models/contract.py]: > Why returning only one record?
When this method is called in the method "recurring_create_invoice" it shows singletone error if multiple invoices are passed.

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld Could you please review this PR?

**Comment by dantiller2** [comment]
What would need to be done to keep this open?

**Comment by dsolanki-initos** [comment]
@ThomasBinsfeld Can you please review this PR?

File: contract_sale_invoicing/models/contract.py (hunk 2 above)
Comment (rousseldenis) [reply contract_sale_invoicing/models/contract.py]: @dsolanki-initos Could you check this comment and find why this change is needed as it is the contrary of the previous code ?

File: contract_sale_invoicing/models/contract.py (hunk 4 above)
Comment (rousseldenis) [reply contract_sale_invoicing/models/contract.py]: Simply don't change the previous code.

It returned a recordset with all generated invoices.

File: contract_sale_invoicing/tests/test_contract_sale_invoicing.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 # Copyright 2018 Tecnativa - Carlos Dauden
+# Copyright 2021 initos Gmbh
```
Comment (rousseldenis) [inline contract_sale_invoicing/tests/test_contract_sale_invoicing.py]: You didn't change the test code, so you don't deserve authorship. Please remove this line

File: contract_sale_invoicing/models/contract.py [hunk 6]
```diff
@@ -1,7 +1,8 @@
 # Copyright 2018 Tecnativa - Carlos Dauden
+# Copyright 2021 initOS Gmbh
```
Comment (rousseldenis) [inline contract_sale_invoicing/models/contract.py]: It's a migration and you don't change so much code. So, please remove this line

**Comment by rousseldenis** [comment]
@dsolanki-initos Please attend comments

**Comment by dsolanki-initos** [comment]
> @dsolanki-initos Please attend comments

Okay I'll change this

File: contract_sale_invoicing/models/contract.py (hunk 1 above)
Comment (rousseldenis) [reply contract_sale_invoicing/models/contract.py]: > Keep simplified syntax. with `super()`.

@dsolanki-initos could you attend this?

**Comment by rousseldenis** [comment]
@dsolanki-initos Could you check tests ?

**Comment by dsolanki-initos** [comment]
> @dsolanki-initos Could you check tests?

@rousseldenis could you please check and add your review?

File: contract_sale_invoicing/tests/test_contract_sale_invoicing.py [hunk 7]
```diff
@@ -45,7 +45,7 @@ def test_sale_invoicing(self):
         self.contract.invoicing_sales = True
         self.sale_order.action_confirm()
         self.contract.recurring_create_invoice()
-        self.assertEqual(self.sale_order.invoice_status, "invoiced")
+        self.assertEqual(self.sale_order.invoice_status, "to invoice")
```
Comment (rousseldenis) [inline contract_sale_invoicing/tests/test_contract_sale_invoicing.py:48]: @dsolanki-initos Could you check why this change ?

Comment (dsolanki-initos) [reply contract_sale_invoicing/tests/test_contract_sale_invoicing.py:48]: This module is migrated from v13.0 and I've checked the test case in v13.0 in the local environment and it also fails

**Review by ikapasi-initos — COMMENTED** [review]
Code review and functional test
LGTM
